extension/chrome/elements/compose-modules/compose-recipients-module.ts

```diff
@@ -798,58 +807,71 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
-    if (contact === PUBKEY_LOOKUP_RESULT_FAIL) {
+    if (sortedPubKeyInfos === PUBKEY_LOOKUP_RESULT_FAIL) {
       recipient.status = RecipientStatus.FAILED;
       $(el).attr('title', 'Failed to load, click to retry');
       $(el).addClass("failed");
       Xss.sanitizeReplace($(el).children('img:visible'), '<img src="/img/svgs/repeat-icon.svg" class="repeat-icon action_retry_pubkey_fetch">' +
         '<img src="/img/svgs/close-icon-black.svg" class="close-icon-black svg remove-reciepient">');
       $(el).find('.action_retry_pubkey_fetch').click(this.view.setHandler(async () => await this.refreshRecipients(), this.view.errModule.handle('refresh recipient')));
       $(el).find('.remove-reciepient').click(this.view.setHandler(element => this.removeRecipient(element.parentElement!), this.view.errModule.handle('remove recipient')));
-    } else if (contact === PUBKEY_LOOKUP_RESULT_WRONG) {
+    } else if (sortedPubKeyInfos === PUBKEY_LOOKUP_RESULT_WRONG) {
       recipient.status = RecipientStatus.WRONG;
       this.view.errModule.debug(`renderPubkeyResult: Setting email to wrong / misspelled in harsh mode: ${recipient.email}`);
       $(el).attr('title', 'This email address looks misspelled. Please try again.');
       $(el).addClass("wrong");
-    } else if (contact.pubkey && ((contact.expiresOn || Infinity) <= Date.now() || contact.pubkey.usableForEncryptionButExpired)) {
-      recipient.status = RecipientStatus.EXPIRED;
-      $(el).addClass("expired");
-      Xss.sanitizePrepend(el, '<img src="/img/svgs/expired-timer.svg" class="revoked-or-expired">');
-      $(el).attr('title', 'Does use encryption but their public key is expired. You should ask them to send ' +
-        'you an updated public key.' + this.recipientKeyIdText(contact));
-    } else if (contact.revoked) {
-      recipient.status = RecipientStatus.REVOKED;
-      $(el).addClass("revoked");
-      Xss.sanitizePrepend(el, '<img src="/img/svgs/revoked.svg" class="revoked-or-expired">');
-      $(el).attr('title', 'Does use encryption but their public key is revoked. You should ask them to send ' +
-        'you an updated public key.' + this.recipientKeyIdText(contact));
-    } else if (contact.pubkey) {
-      recipient.status = RecipientStatus.HAS_PGP;
-      $(el).addClass('has_pgp');
-      Xss.sanitizePrepend(el, '<img class="lock-icon" src="/img/svgs/locked-icon.svg" />');
-      $(el).attr('title', 'Does use encryption' + this.recipientKeyIdText(contact));
+    } else if (sortedPubKeyInfos.length) {
+      // New logic
+      // - if there is at least one valid (non-expired, non-revoked) public key, then it's HAS_PGP
+      // - else if there is at least one expired public key, then it's EXPIRED
+      // - else if there is at least one revoked key, then REVOKED
+      // - else it's NO_PGP
+      const firstKeyInfo = sortedPubKeyInfos[0];
+      if (!firstKeyInfo.revoked && !PubKeyInfoUtil.isExpired(firstKeyInfo)) {
```

Comment:
Actually, a valid key is a key that has `!firstKeyInfo.revoked && firstKeyInfo.pubkey.usableForEncryption` (there was already time check at the moment of `Key` object creation).
Ok, if you want to double-check if the key hasn't expired within seconds (minutes) after retrieval, you can additionally `&& !KeyUtil.expired(firstKeyInfo.pubkey)`